### ISSUE TITLE
The transition animation doesn't work for the request money page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#22388]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Click on the fab button and select request money
2. Enter an amount
3. Click on the currency or on next
4. Watch what happens
## Expected Result: 
The new page opens with an horizontal animation
## Actual Result: 
The new page opens without animation

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop
 
**Version Number:** 1.3.37-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/abb8db8a-e2ed-40d1-8338-f69ce21f64a5

**Expensify/Expensify Issue URL:**
**Issue reported by:** 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688606747153309

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When we click on the currency or next in requestmoney page there is no horizontal sliding animation

### What is the root cause of that problem?

The root cause of that problem is that in:
- IOUCurrencySelection
- MoneyRequestParticipantsSelector
- MoneyRequestParticipantsSplitSelector

 we use an OptionsSelector for example here https://github.com/Expensify/App/blob/c0ece67e0edcb725d8f72cf57cd40d98854c552c/src/pages/iou/IOUCurrencySelection.js#L119-L129

By default OptionSelectors focus the input here:
https://github.com/Expensify/App/blob/c0ece67e0edcb725d8f72cf57cd40d98854c552c/src/components/OptionsSelector/BaseOptionsSelector.js#L122

And that can cause the transition animation to be stopped.

### What changes do you think we should make in order to solve the problem?


We should pass the prop
`shouldDelayFocus={true}`
to the OptionSelector so that the input waits for the animation to finish before focusing the input


 Before this change:

https://github.com/Expensify/App/assets/19537677/5bbd9594-34aa-4c40-a0ba-5aa70f09126b

After this change:

https://github.com/Expensify/App/assets/19537677/dfc4b023-19b8-4b67-9110-e3f9fd662f9c


### What alternative solutions did you explore? (Optional)

**Comment by ShogunFire** [comment]
@chiragxarora I am not sure but maybe that has to do with that PR where we added withNavigationFocus: https://github.com/Expensify/App/pull/19824/files

**Comment by JmillsExpensify** [comment]
Hmm, yeah I'm not sure how/why these pages stopped being animated with the page coming in from the right. Regardless, I agree we should consistently animate, so I'm opening this one up to the community.

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The transition animation doesn't work for the request money page 

### What is the root cause of that problem?
Root Cause of the issue is that we have an input inside the modal which is having `autoFocus` prop default to true which is causing and breaking the animation. Ideally we should only focus the input or render anything at all inside the modal only after transition done.

Existing issues like #22036 does possess same problem like this.

### What changes do you think we should make in order to solve the problem?

The simple solution we can use is adding `shouldDelayFocus` but according to the other issues I followed up `shouldDelayFOcus` can't be used because of `setTimeOut` used inside the logic. Even more to this if the list of user-contacts is high (if it's high traffic account) this issue comes again.

To resolve this we should manually handle focus of input.

## Changes
https://github.com/Expensify/App/blob/a047b88851917807b036f3c5b8564c5add9ea9d2/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSplitSelector.js#L167 Here we should add `autoFocus={false}` to disable auto-focus.

From here
https://github.com/Expensify/App/blob/a047b88851917807b036f3c5b8564c5add9ea9d2/src/pages/iou/steps/MoneyRequestAmountPage.js#L415

We can take the state `didScreenTransitionEnd` and pass it to the https://github.com/Expensify/App/blob/a81b63ac88bc35268e278b402e4bfb0490ba3eec/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js#L98 here.

Like the following
```jsx
autoFocus={didScreenTransitionEnd}
```

Using this inside these two components

https://github.com/Expensify/App/blob/a81b63ac88bc35268e278b402e4bfb0490ba3eec/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js#L106

https://github.com/Expensify/App/blob/a81b63ac88bc35268e278b402e4bfb0490ba3eec/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js#L98

We create ref maybe like 
```jsx
const optionRef = useRef(null)
```
and assign this ref here 
https://github.com/Expensify/App/blob/a81b63ac88bc35268e278b402e4bfb0490ba3eec/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSelector.js#L161

using the below code

```jsx
ref={(element) => optionRef.current = element}
```

And when we see the `props.autoFocus` as true within `componentDidUpdate` or `useEffect` we do autoFocus using

```jsx
optionRef.current.textInput.focus()
```

Also if we have a high traffic account which does have multiple user contacts in that case also the transition could get effected to fix this we should add a check here as well

https://github.com/Expensify/App/blob/a81b63ac88bc35268e278b402e4bfb0490ba3eec/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSelector.js#L158 To check weather transition ended or not then only we do show options until then loader will be shown.


### What alternative solutions did you explore? (Optional)

I have seen this issue happens in other places as well, we could consider conditional rendering here
https://github.com/Expensify/App/blob/85739ea075b554a5a15c8750f2d92bed27005f2f/src/components/ScreenWrapper/index.js#L84

Adding a conditional rendering based on the state `didScreenTransitionEnd` could fix this issue at all maybe.

**Comment by b4s36t4** [comment]
It was working for me.

**Comment by b4s36t4** [comment]
@chiragxarora Updated the proposal with the video.

**Comment by ShogunFire** [comment]
I get that you guys don't like timeout but I don't see why we are changing something that is working everywhere without problem ... Anyway you are the boss.

**Comment by ShogunFire** [comment]
Btw if I understand correctly someone is trying to fix this problem in every page at the same time here: https://github.com/Expensify/App/issues/19530#issuecomment-1605291482

**Comment by eVoloshchak** [comment]
Not overdue, reviewing this tomorrow

**Comment by eVoloshchak** [comment]
Not overdue, reviewing this tomorrow

**Comment by eVoloshchak** [comment]
@tienifr's proposal in https://github.com/Expensify/App/issues/22271#issuecomment-1622782156 has already resolved this for another page, let's use the same approach

>So far I have found three affected pages:

@tienifr, were you able to find other affected pages?

**Comment by eVoloshchak** [comment]
> Since we're doing this on all the pages, does it make sense to find a fix for the reload problem as well?

I think yes, it's definitely something worth fixing
Is it reproducible on all platforms? 
I remember there's a similar issue specific to Safari

**Comment by JmillsExpensify** [comment]
I can get this on iOS Safari. What are the next steps?

**Comment by ShogunFire** [comment]
Sorry to beat a dead horse but with shouldDelayFocus I don't think there was a problem during reload, I still don't really understand why we decided to change a method that was working

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/22388#issuecomment-1624482708)

**Comment by ShogunFire** [comment]
I might have found a good way with  InteractionManager.runAfterInteractions . I will test it more because I am not sure but it seems to work on android chrome

**Comment by eVoloshchak** [comment]
Reviewing updated proposal tomorrow

**Comment by eVoloshchak** [comment]
@ShogunFire, thanks for the updated [proposal](https://github.com/Expensify/App/issues/22388#issuecomment-1624482708)
I don't think we should use `InteractionManager.runAfterInteractions`. There is no hard guarantee that it will be fired upon screen transition end on all platforms. In addition to that, it will be fired after all animations have completed, not the specific screen transition animation
`didScreenTransitionEnd` is exactly what we need and what has been used to fix this on other pages (for instance, https://github.com/Expensify/App/pull/22217)

Looks like https://github.com/Expensify/App/issues/19530 already has a proposal to resolve this universally throughout the app, I think we should put this on hold/combine those issues

**Comment by JmillsExpensify** [comment]
Still working through proposals.

**Comment by eVoloshchak** [comment]
Not overdue, waiting for more proposals
> Looks like https://github.com/Expensify/App/issues/19530 already has a proposal to resolve this universally throughout the app, I think we should put this on hold/combine those issues

That issue ended up applying a [non-general fix](https://github.com/Expensify/App/issues/19530#issuecomment-1667003618), so disregard the part about putting this one on HOLD

> Given that this is not critical issue which affects user, non-general fix is fine to me. If anyone concerns, general solution can still be researched and discussed in thread where @dukenv0307 reported [here](https://expensify.slack.com/archives/C049HHMV9SM/p1690515295054879).

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/22388#issuecomment-1624482708)

To be honest I haven't tried this solution, but do you think it could be the right way ?

**Comment by JmillsExpensify** [comment]
Working through proposals

**Comment by eVoloshchak** [comment]
@tienifr's [proposal](https://github.com/Expensify/App/issues/22388#issuecomment-1672397143) looks good to me!
It's universal, there are a couple of already open issues that would be resolved by this fix (https://github.com/Expensify/App/issues/24545 for instance, we should add `Pronous selector page` to the list of pages needed to fix)

As for the third item, we should use the following: `we can expose the ref of the text input out to the MoneyRequestParticipantsPage and focus the input by using onEntryTransitionEnd`

We also have to make sure to add this to docs/reviewer checklist I think, so all the newly created pages don't have the same issue

🎀👀🎀 C+ reviewed!

**Comment by eVoloshchak** [comment]
@b4s36t4, apologies for the confusion
@tienifr has proposed this approach on Jul 6 [here](https://github.com/Expensify/App/issues/22271#issuecomment-1622782156), but linked that proposal [here](https://github.com/Expensify/App/issues/22388#issuecomment-1632989168) on Jul 12 (after you've posted your proposal with the same approach)

So while this solution was originally proposed by @tienifr in another issue, @b4s36t4's proposal was technically the first valid proposal for this issue
The rules are the rules, I think it's fair if we proceed with @b4s36t4's proposal
cc: @roryabraham

**Comment by b4s36t4** [comment]
@ShogunFire Refresh on focus is different problem it happens because of the `transitionEnd` event which we cannot do anything about it though.

Thanks @eVoloshchak I didn't know we were avoiding the `shouldDelayFocus` so I've taken inspiration, this issue is my 3rd (1st and 2nd were also same issues).

**Comment by b4s36t4** [comment]
@ShogunFire more on this here https://github.com/Expensify/App/issues/22850#issuecomment-1686459380

**Comment by ShogunFire** [comment]
Refresh on focus is not a different problem since this is a regression caused by the solution we are all proposing, I think we should find a solution to this regression before assigning anyone @eVoloshchak , I thought @tienifr proposal was solving it but if it's not the case I proposed two other ways to solve it in my proposal 


> In ScreenWrapper before we declare the listener for the end of the transition we can check the active route index in the stack, if it is 0 it means that we accessed this page without transition and we can call onEntryTransitionEnd directly without passing by the listener

and

> We can set up a boolean isTransitionHappening to true by listening to the event transitionStart, I think we should do that in Navigation. And test if that boolean is true before setting up the listener for the end of the transition

**Comment by b4s36t4** [comment]
@tienifr I think you haven't read my proposal clearly please read it and watch the demo I have shared. Thanks.

**Comment by b4s36t4** [comment]
Yes, instead of exposing the ref from Inner components to the screen I'm passing the transition state to the inner components from screen and inside the components using the ref to focus based on the prop passed from the screen.

**Comment by JmillsExpensify** [comment]
I lost the thread on this issue. Someone mind linking the PR that's being worked on?

**Comment by ShogunFire** [comment]
@JmillsExpensify , @eVoloshchak has reviewed but nobody got assigned yet, I was hoping to change his mind about approving any proposal that doesn't fix the refresh problem since he said this before: https://github.com/Expensify/App/issues/22388#issuecomment-1641037977

**Comment by eVoloshchak** [comment]
Reviewing shortly

**Comment by b4s36t4** [comment]
@eVoloshchak As far as I remember this refresh problem is not re-produceable on prod/staging because of bundle size and how the event fires.  And it's not only related to this issue it's the issue why `transitionEnd` is not firing when we do refresh. 

I think all we have to do fix the main transition problem/ for focusing I think we have other issue #22850. Please take a look here.

**Comment by eVoloshchak** [comment]
> As far as I remember this refresh problem is not re-produceable on prod/staging because of bundle size and how the event fires

@b4s36t4, It's still present in DEV. We should fix it, we should strive for zero bugs, even if this isn't reproducible in prod, it can be in the future if we don't resolve it and the bundle size changes

@JmillsExpensify, I think we should proceed with @tienifr's [proposal](https://github.com/Expensify/App/issues/22388#issuecomment-1672397143)
It's substantially different with the usage of `onEntryTransitionEnd`, but mainly this came down to the fact that we do need to fix the refresh regression as part of this PR

**Comment by b4s36t4** [comment]
@eVoloshchak I don't think we're solving the correct problem here nor we have the correct RCA. Even though we want to solve the refresh problem it's neither related to any PR because we don't want to use `shouldDelayFocus` we're depending on `transitionEnd` event. But all we're trying to find an alternative to make `didScreenTranstionEnd` as true instead of solving the real issue why there's is no `transitionEnd` event firing. I'd say this is pretty much troublesome approach which might cause issues later. We're trying to fix it by over-engineering the solution. Maybe I'm wrong but this is what I'm feeling about the solutions.

**Comment by roryabraham** [comment]
90% sure this will be fixed by https://github.com/Expensify/App/pull/26415

**Comment by roryabraham** [comment]
HOLDing because I think it will be fixed by https://github.com/Expensify/App/pull/26415, which I will be able to merge as soon as we lock the new checklist today

**Comment by roryabraham** [comment]
https://github.com/Expensify/App/pull/26415 is on prod, retesting

**Comment by roryabraham** [comment]
Verified this is fixed on staging, but not yet on prod. Maybe something weird with deploy comments. I'll check back on Tuesday and see if it's still reproducible.

**Comment by eVoloshchak** [comment]
This was resolved for the case when you press 'Save', but still is present when you click on the amount to edit it

https://github.com/Expensify/App/assets/9059945/197a7030-a5bc-418c-abe0-a6f64e15e668

**Comment by ShogunFire** [comment]
Hello, I accepted the offer but the reporting bounty was 250 when this issue arrived in github (3months ago)

**Comment by roryabraham** [comment]
@JmillsExpensify this issue appears to be eligible for the #urgency bonus. The PR was created, tested, and C+ approved within 3 working days of assignment. It was only delayed by me, so I think it's fair that @tienifr and @eVoloshchak get the urgency bonus. Thanks!

**Comment by mountiny** [comment]
@JmillsExpensify There was a regression from this PR so not urgency bonus and we should recude the price as the fix in sake of urgency is handled by other available engineers https://github.com/Expensify/App/pull/28557

**Comment by mountiny** [comment]
oops, yes, but speaking of devil this deploy blocker seems highly related to this PR/issue https://github.com/Expensify/App/issues/28523 can you have a look @tienifr @eVoloshchak

**Comment by mountiny** [comment]
Coming back from this issue https://github.com/Expensify/App/issues/28523 the PR linked is already in staging, but the request money page transitions are not working. also noticed different bug which I will report in expensify bugs

https://github.com/Expensify/App/assets/36083550/28084617-0ba5-4025-b044-c81c611066e0

**Comment by mountiny** [comment]
@tienifr for example on the 14th second you can see the page blinks completely, in others you can see the transition is not smooth

I understand if that is related to more generic issues such as too heavy rerenders, but wanted to raise that

**Comment by JmillsExpensify** [comment]
I believe based on the above the correct payment summary is:

 - Issue reporter: @ShogunFire $250 (reported before 25 Aug)
 - Contributor: @tienifr $1,500 (incl. [urgency bonus](https://github.com/Expensify/App/issues/22388#issuecomment-1740179636))
 - Contributor+: @eVoloshchak $1,500 (incl. urgency bonus)

@mountiny Did you potentially want to ask for a follow-up PR, or are you good with this summary?

**Comment by mountiny** [comment]
We are good here

**Comment by JmillsExpensify** [comment]
Issue reporter and contributor have been paid. @eVoloshchak Please complete the BZ checklist and add a request via NewDot so we can get this issue closed. Thanks!

**Comment by JmillsExpensify** [comment]
$1,500 payment approved for @eVoloshchak based on BZ summary.
